cmd/server/apiserver/argoserver.go

```diff
@@ -0,0 +1,271 @@
+	"github.com/argoproj/argo/pkg/client/clientset/versioned"
+	grpcutil "github.com/argoproj/argo/util/grpc"
+	"github.com/argoproj/argo/util/json"
+	"github.com/argoproj/argo/workflow/common"
+	"github.com/argoproj/argo/workflow/config"
+)
+
+type ArgoServer struct {
+	namespace        string
+	kubeClientset    *kubernetes.Clientset
+	wfClientSet      *versioned.Clientset
+	enableClientAuth bool
+	insecure         bool
+	configName       string
+	stopCh           chan struct{}
+}
+
+type ArgoServerOpts struct {
+	Insecure         bool
+	Namespace        string
+	KubeClientset    *kubernetes.Clientset
+	WfClientSet      *versioned.Clientset
+	EnableClientAuth bool
+	ConfigName       string
+}
+
+func NewArgoServer(opts ArgoServerOpts) *ArgoServer {
+	return &ArgoServer{
+		namespace:        opts.Namespace,
+		wfClientSet:      opts.WfClientSet,
+		kubeClientset:    opts.KubeClientset,
+		enableClientAuth: opts.EnableClientAuth,
+		insecure:         opts.Insecure,
+		configName:       opts.ConfigName,
+	}
+}
+
+var backoff = wait.Backoff{
+	Steps:    5,
+	Duration: 500 * time.Millisecond,
+	Factor:   1.0,
+	Jitter:   0.1,
+}
+
+func (as *ArgoServer) useTLS() bool {
+	return !as.insecure
+}
+
+func (as *ArgoServer) Run(ctx context.Context, port int) {
+	grpcServer := as.newGRPCServer()
+	var httpServer *http.Server
+	var httpsServer *http.Server
+	if as.useTLS() {
+		httpServer = newRedirectServer(port)
+		httpsServer = as.newHTTPServer(ctx, port)
+	} else {
+		httpServer = as.newHTTPServer(ctx, port)
+	}
+
+	// Start listener
+	var conn net.Listener
+	var listerErr error
+	err := wait.ExponentialBackoff(backoff, func() (bool, error) {
+		conn, listerErr = net.Listen("tcp", fmt.Sprintf("localhost:%d", port))
+		if listerErr != nil {
+			log.Warnf("failed to listen: %v", listerErr)
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	// Cmux is used to support servicing gRPC and HTTP1.1+JSON on the same port
+	tcpm := cmux.New(conn)
+	var tlsm cmux.CMux
+	var grpcL net.Listener
+	var httpL net.Listener
+	var httpsL net.Listener
+	if !as.useTLS() {
+		httpL = tcpm.Match(cmux.HTTP1Fast())
+		grpcL = tcpm.Match(cmux.Any())
+	} else {
+
+		// If not matched, we assume that its TLS.
+		tlsl := tcpm.Match(cmux.Any())
+		tlsConfig := tls.Config{
+			//Certificates: []tls.Certificate{*as.settings.Certificate},
+		}
+
+		tlsl = tls.NewListener(tlsl, &tlsConfig)
+
+		// Now, we build another mux recursively to match HTTPS and gRPC.
+		tlsm := cmux.New(tlsl)
+		httpsL = tlsm.Match(cmux.HTTP1Fast())
+		grpcL = tlsm.Match(cmux.Any())
+	}
+
+	go func() { as.checkServeErr("grpcServer", grpcServer.Serve(grpcL)) }()
+	go func() { as.checkServeErr("httpServer", httpServer.Serve(httpL)) }()
+	go func() { as.checkServeErr("tcpm", tcpm.Serve()) }()
+	if as.useTLS() {
+		go func() { as.checkServeErr("httpsServer", httpsServer.Serve(httpsL)) }()
+		go func() { as.checkServeErr("tlsm", tlsm.Serve()) }()
```

Comment:
This line is a nil deref when not using TLS (`tlsm` is not initialized):

```
time="2019-12-19T17:33:47Z" level=info msg="Config Map: metricsConfig:\n  enabled: true\n  path: \"/metrics\"\n  port: 9091\n"
time="2019-12-19T17:33:47Z" level=info msg="Argo Server started successfully on port 2746"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x152e022]

goroutine 49 [running]:
github.com/argoproj/argo/cmd/server/apiserver.(*ArgoServer).Run.func6(0xc000430ac0, 0x0, 0x0)
	~/go/src/github.com/argoproj/argo/cmd/server/apiserver/argoserver.go:133 +0x22
created by github.com/argoproj/argo/cmd/server/apiserver.(*ArgoServer).Run
	~/go/src/github.com/argoproj/argo/cmd/server/apiserver/argoserver.go:133 +0x5e8
```

Reply:
this is a bug - but I think we plan to remove SSL support